bin/warnhist

```diff
@@ -67,29 +86,196 @@ sub do_help {
+
+    # Attempt to detect and handle object file name mangling by Automake
+    if(!exists $file_paths->{$filename} && $filename =~ /\-/) {
+        my ($bn) = $filename =~ /\S+\-(\S+)$/x;
+        if(exists $file_paths->{$bn}) {
+            if(!exists $options{q}) {
+                warn "No path for $typename source file '$filename', but '$bn' has path, assuming automake generated object file name";
+            }
+            $filename = $bn;
+        }
+    }
+
+    if(exists $file_paths->{$filename}) {
+        my $filecount = 0;
+
+        # Attempt to count LOC for files with same name
+        if(scalar(@{$file_paths->{$filename}}) > 1) {
+            $filecount = $files_counted->{$filename}++;
+
+            # Issue warning about multiple source files with same name
+            if($filecount == 0 && !exists $options{q}) {
+                local $" = ', ';        # '$"' is documented in https://perldoc.perl.org/perlvar
+                warn "Multiple paths for $typename source file named '$filename', assuming each is compiled once, paths: [@{$file_paths->{$filename}}]\n";
+            }
+
+            # Sanity check for too many compiles of a file
+            if($filecount >= scalar(@{$file_paths->{$filename}})) {
+                if(!exists $options{q}) {
+                    local $" = ', ';        # '$"' is documented in https://perldoc.perl.org/perlvar
+                    warn "Too many compiles of $typename source file named '$filename' with paths: [@{$file_paths->{$filename}}], disabling warning density calculations\n";
+                }
+                delete $options{d};
+            }
+        }
+
+        # Increment the # of lines of code (if not too many)
+        if($filecount < scalar(@{$file_paths->{$filename}})) {
+            ${$count} += line_count($file_paths->{$filename}[$filecount]);
+        }
+    } else {
+        if(!exists $options{q}) {
+            warn "No path for $typename source file '$filename', e '-p' option to specify, disabling warning density calculations\n";
+        }
+        delete $options{d};
+    }
+}
+
+# Compute LOC for compiled source file
+sub count_source_loc {
+    my ($compile_line) = @_;
+#print "compile_line = $compile_line\n";
+    my $filetype;
+    my $filename;
+
+    ($filetype, $filename) = $compile_line =~ /^\s+(CC|FC|CXX|PPFC)\s+(\S*)\.l*o$/x;
+    if($filename =~ /\//) {
+        $filename = basename($filename);
+    }
+#print "filetype = '$filetype'\n";
+#print "filename = '$filename'\n";
+
+    if($filetype =~ /FC|PPFC/) {        # FORTRAN source file
+        count_file_loc($filename, "FORTRAN", \%fort_files, \%fort_files_counted, \$fort_lines);
+    } elsif($filetype =~ /CXX/) {       # C++ source file
+        count_file_loc($filename, "C++", \%cpp_files, \%cpp_files_counted, \$cpp_lines);
+    } elsif($filetype =~ /CC/) {        # C source file
+        count_file_loc($filename, "C", \%c_files, \%c_files_counted, \$c_lines);
+    }
+}
+
+sub sanity_check_loc {
```

Comment:
Make certain that the number of compiles of a source file name equals the number of files with that name.